### PR TITLE
Add 404 page

### LINKED ISSUE #474: Set up custom 404 page
**Describe the solution you'd like**
This page should ideally show some clever message and point users back to the homepage. [See this article](https://designshack.net/articles/inspiration/what-makes-a-great-404-error-page/) for 404 page best practices.

**Done when**
Any mistyped URL takes users to a 404 page, which redirects them to the homepage or other important pages. Don't worry about trying to fix typos in the URL or anything.

### PR DESCRIPTION
**Pull request by syntapy**
### Description
This will add a custom 404 page and enable the default error handler to render it

404pg_mobile_es

404pg_mobile

404pg_desktop_es

404pg_desktop

- [x] Make sure the error handler is triggered correctly in dev
- [x] Fix slight spacing issue with top right login button
- [x] Fix slight spacing issue with mobile version of header / description text
- [x] Extend the browser test to make sure the login button displays correctly under different states
- [x] Use a better login icon image that doesn't blur at higher zoom levels (maybe a vector graphics or CSS to do it?)
- [x] Upload pictures of 404 page to this PR
- [x] Add spanish translation (rest will be added later)

### Issue(s)
Fixes #474

### CODE REVIEW

File: universal-application-tool-0.0.1/conf/messages.en-US [hunk 1]
```diff
@@ -45,6 +45,19 @@ content.benefits=Get benefits
 content.description1=CiviForm lets you apply for many benefits at once by reusing information.
 content.description2=Get started by choosing an application below.
 
+error.not_found_title=We''re sorry, we lost
+error.not_found_title_end=the page
+error.not_found_description_a=Don''t worry though, the benefits
+error.not_found_description_a_end=are still there.
+error.not_found_description_b=Find them on
+error.not_found_description_b_end=the
+error.not_found_description_link=homepage
```
Comment (bion) [inline universal-application-tool-0.0.1/conf/messages.en-US:54]: issue (blocking): please reword to "We were unable to find the page you tried to visit. Please go back or return to the homepage."

File: universal-application-tool-0.0.1/conf/messages.en-US [hunk 2]
```diff
@@ -45,6 +45,19 @@ content.benefits=Get benefits
...
+
+error.img_caption_a=Photo by
+error.img_caption_b=Michael Dziedzic
+error.img_caption_c=on
+error.img_caption_d=Unsplash
+
```
Comment (bion) [inline universal-application-tool-0.0.1/conf/messages.en-US:60]: issue (blocking): please remove the unsplash image

**Review by bion — CHANGES_REQUESTED** [review]
@syntapy sorry for taking so long to get around to looking at this, and thank you for doing it! Just a few things and it will be ready to merge:

- [ ] remove the unsplash image
- [ ] remove the Seattle logo
- [ ] reword the error text as mentioned in the comment

File: server/app/services/applicant/ApplicantData.java [hunk 3]
```diff
@@ -79,7 +79,7 @@ public Optional<String> getApplicantName() {
     } catch (NoSuchElementException e) {
       logger.warn(
           "Application {} does not include an applicant name. This is expected for guest users.");
-      return Optional.empty();
+      return Optional.of("Guest");
```
Comment (syntapy) [inline server/app/services/applicant/ApplicantData.java:82]: Need to revisit, since I think it was still displaying "Logged in as Guest" in the nav bar before

File: server/conf/messages.en-US [hunk 4]
```diff
@@ -340,3 +343,12 @@ validation.textTooShort=Must contain at least {0} characters.
 
 # Validation errors that appear if an admin leaves a multi option question blank.
 adminValidation.multiOptionEmpty=Multi option questions cannot have blank options.
+
+#-------------------------------------------------------#
```
Comment (gwendolyngoetz) [inline server/conf/messages.en-US:347]: The same changes in the messages.en-US file should also be done to the messages files.

File: server/app/views/BaseHtmlView.java [hunk 5]
```diff
@@ -79,6 +81,22 @@ protected static ButtonTag makeSvgTextButton(String buttonText, Icons icon) {
             span(buttonText).withClass(Styles.TEXT_LEFT));
   }
 
+  protected static SpanTag space() {
+    return span(" ");
```
Comment (gwendolyngoetz) [inline server/app/views/BaseHtmlView.java:85]: Is `" "` ok in this case or should it be a `&nbsp;`

Comment (syntapy) [reply server/app/views/BaseHtmlView.java:85]: Well it is true, that is under the assumption of latin-based languages. I guess there at most 4 languages civiform supports / will support that might not have the same character for spaces....

Maybe I should put the space in the messages file instead?

Comment (syntapy) [reply server/app/views/BaseHtmlView.java]: Done

File: server/conf/messages.en-US (hunk 4 above)
Comment (syntapy) [reply server/conf/messages.en-US:351]: I tried it. Did I do it right?

File: server/app/views/BaseHtmlView.java (hunk 5 above)
Comment (syntapy) [reply server/app/views/BaseHtmlView.java]: Done. Hows this now?

File: server/app/views/errors/NotFound.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  public NotFound(ApplicantLayout layout, LanguageSelector languageSelector) {
+    this.layout = layout;
+    this.languageSelector = checkNotNull(languageSelector);
+  }
+
+  private H1Tag h1Content(Messages messages) {
+    return h1(
+            span(messages.at(MessageKey.ERROR_NOT_FOUND_TITLE.getKeyName())),
+            span(messages.at(MessageKey.GENERAL_SPACE.getKeyName())),
+            spanNowrap(messages.at(MessageKey.ERROR_NOT_FOUND_TITLE_END.getKeyName())))
```
Comment (bion) [inline server/app/views/errors/NotFound.java:40]: issue (blocking): instead of breaking this into multiple messages, create a single message that takes one or more arguments: https://www.playframework.com/documentation/2.8.x/JavaI18N#Formatting-messages

File: server/app/views/errors/NotFound.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import views.style.Styles;
+
+public class NotFound extends BaseHtmlView {
+
+  private final ApplicantLayout layout;
+  private final LanguageSelector languageSelector;
+
+  @Inject
+  public NotFound(ApplicantLayout layout, LanguageSelector languageSelector) {
+    this.layout = layout;
```
Comment (bion) [inline server/app/views/errors/NotFound.java:32]: nit: checkNotNull

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/views/BaseHtmlView.java (hunk 5 above)
Comment (gwendolyngoetz) [reply server/app/views/BaseHtmlView.java]: I'm not sure about how other languages would use different space characters. In general I'd just say keep it as part of the full text being translated rather than having a separate character as it's own thing.

Comment (syntapy) [reply server/app/views/BaseHtmlView.java]: yeah, back then I didn't know how to put a space at the end of the message entry, which is why I did that I think...

File: server/app/views/errors/NotFound.java (hunk 6 above)
Comment (syntapy) [reply server/app/views/errors/NotFound.java]: I just got rid of the spanNowrap call. But I left the method in there in case someone else needs it. Otherwise I think its ready for another look

File: server/app/controllers/ErrorHandler.java [hunk 8]
```diff
@@ -51,8 +56,12 @@ public ErrorHandler(
       Config config,
       Environment environment,
       OptionalSourceMapper sourceMapper,
-      Provider<Router> routes) {
+      Provider<Router> routes,
+      NotFound notFoundPage,
+      MessagesApi messagesApi) {
     super(config, environment, sourceMapper, routes);
+    this.notFoundPage = notFoundPage;
+    this.messagesApi = messagesApi;
```
Comment (bion) [inline server/app/controllers/ErrorHandler.java:64]: suggestion (blocking): `checkNotNull`

File: server/app/controllers/ErrorHandler.java [hunk 9]
```diff
@@ -100,4 +109,10 @@ static Optional<Throwable> findThrowableByTypes(
     }
     return Optional.empty();
   }
+
+  @Override
+  public CompletionStage<Result> onNotFound(RequestHeader request, String message) {
+    return CompletableFuture.completedFuture(
+        Results.ok(notFoundPage.render(request, messagesApi.preferred(request))));
```
Comment (bion) [inline server/app/controllers/ErrorHandler.java:116]: suggestion (blocking): the response code for not found should be 404, please change to `Results.notFound(...)`

File: server/app/views/errors/NotFound.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+import services.MessageKey;
+import views.BaseHtmlView;
+import views.HtmlBundle;
+import views.LanguageSelector;
+import views.applicant.ApplicantLayout;
+import views.style.BaseStyles;
+import views.style.ErrorStyles;
+import views.style.Styles;
+
+public class NotFound extends BaseHtmlView {
```
Comment (bion) [inline server/app/views/errors/NotFound.java:25]: suggestion (blocking): add `final`

File: server/app/views/errors/NotFound.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    bundle.setLanguage(language);
+    bundle.addMainContent(mainContent(messages));
+
+    return bundle;
+  }
+
+  public Content render(Http.RequestHeader request, Messages messages) {
+    HtmlBundle bundle = addBodyFooter(request, messages);
+    return layout.render(bundle);
+  }
```
Comment (bion) [inline server/app/views/errors/NotFound.java:69]: suggestion (blocking): please move the public method above the private methods

File: server/conf/messages.en-US [hunk 12]
```diff
@@ -42,6 +42,10 @@ validation.isRequired=This question is required.
 # Validation error that is shown when the user input couldn't be converted for any reason.
 validation.invalidInput=Please enter valid input.
 
+# For use in cases where sentences are broken up
+general.space=' '
+general.period=.
+
```
Comment (bion) [inline server/conf/messages.en-US:48]: suggestion (blocking): these aren't being used, please remove here and other references

File: server/conf/messages.en-US [hunk 13]
```diff
@@ -136,6 +140,9 @@ content.description1=CiviForm lets you apply for many benefits at once by reusin
 # Long form description of the CiviForm site shown to the applicant. Line 2.
 content.description2=Get started by choosing an application below.
 
+label.street=Address line 1
+placeholder.street=Street address
+
```
Comment (bion) [inline server/conf/messages.en-US:145]: suggestion (blocking): these look unrelated to this PR?

File: server/package-lock.json [hunk 14]
```diff
@@ -358,9 +358,9 @@
       }
     },
     "node_modules/caniuse-lite": {
-      "version": "1.0.30001373",
-      "resolved": "https://registry.npmjs.org/caniuse-lite/-/caniuse-lite-1.0.30001373.tgz",
-      "integrity": "sha512-pJYArGHrPp3TUqQzFYRmP/lwJlj8RCbVe3Gd3eJQkAV8SAC6b19XS9BjMvRdvaS8RMkaTN8ZhoHP6S1y8zzwEQ==",
+      "version": "1.0.30001238",
+      "resolved": "https://registry.npmjs.org/caniuse-lite/-/caniuse-lite-1.0.30001238.tgz",
+      "integrity": "sha512-bZGam2MxEt7YNsa2VwshqWQMwrYs5tR5WZQRYSuFxsBQunWjBuXhN4cS9nV5FFb1Z9y+DoQcQ0COyQbv6A+CKw==",
```
Comment (bion) [inline server/package-lock.json:363]: suggestion (blocking): please omit the package-lock.json diff from this PR

**Review by bion — CHANGES_REQUESTED** [review]

**Review by bion — APPROVED** [review]
